fix: +1 for L1 latency and trigger class check improved

lietava opened the pull request:
Two issues:
1.)
+1 for L1 input latency.
L1 latency for Trigger Class and Trigger input differs by 1. It is constant
2.) check if class has non zero inputs
if parallel runs then this check produce errors. Therefore class mask for trigger classes for a given run is checked only

lietava committed:
fix: +1 for L1 latency

lietava committed:
fix: +1 for L1 latency

lietava commented:
async-*-pbpb-*
async-*-PbPb-*

lietava commented:
Hi @shahor02 , for the labels this is needed for all PbPb reco.

shahor02 commented:
Hi @lietava , could you please provide more details? Is this only for the pbpb24 or also for pp and pbpb23? 
We use L0_L1 also here: https://github.com/AliceO2Group/AliceO2/blob/f1eaeb2aeb798b53f924cdd83fbb96c400f6c9f6/Detectors/CTP/reconstruction/src/RawDataDecoder.cxx#L92 
Is it ok?

lietava committed:
dev: decoder: checking only trigger class bits which belongs to run

lietava committed:
clang

lietava commented:
Hi @shahor02 , thanks for question. Please, read description - I updated.
Yes, I am aware that L0-L1 latency is used also on the place you pointed out. That one is used only for trigger classes.
I realised that there is difference 1 BC between trigger class latency and input latency - fw feature.
This latency fix concernes only ZDC inputs - we did not use any other L1 inputs, so it is not critical.
For year 2023 the zdc inputs were properly aligned only in the middle of the run - I would not introduce special fix for that.
For 2024 after fix all inputs should be sitting at LM time.
Then other issue: in decodong inputs (shiftInputs) I am doinf check if trigger class has non zero inputs in same BC. However if there is parallel technical run with internal ctp input generaors it produces a lot of errors. Data are ok , just errors are meningless.
So I introduced fix for that. Disadvantege is that I need to access CTPConfig. If that is a problem I remove error check .
I introduced fix only RawDataDecoder - if you approve this also CTFCoder.h has to access CTPConfig - I did not do that.

shahor02 reviewed (dismissed):
Thanks for explanation, will merge after CI

shahor02 commented on Detectors/CTP/macro/TestConfig.C:
```diff
@@ -16,13 +16,13 @@
 #endif
```
@lietava could you add 
```
#include "Framework/Logger.h"
```
after(!) the the `#endif` to pass macro test in the interpreter mode?

lietava replied:
Ahh, I didnt want to commit macros. Yes, I will do something. But I have other question, please, see in review.

shahor02 replied:
@lietava the `#include "Framework/Logger.h"` should have been added *after* the `#endif`, to be visible in the interpreter mode, which is currently failing.

lietava commented on Detectors/CTP/workflow/src/RawDecoderSpec.cxx:
```diff
@@ -135,6 +137,11 @@ void RawDecoderSpec::run(framework::ProcessingContext& ctx)
     ret = mDecoder.decodeRawFatal(inputs, filter);
   } else {
+    if (mDecodeinputs) {
+      const auto ctpcfg = inputs.get<o2::ctp::CTPConfiguration*>("ctpconfig");
```
what will happen if ctpcfg will not be found ?
nullptr is returned ?

shahor02 replied:
Fatal

lietava committed:
dev: ctp config added also to CTF decoder

lietava committed:
clang

alibuild commented:
Error while checking build/O2/fullCI for 0de363f5596be5488cdad3cd3fa58172dd51c23a at 2024-11-22 09:16:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:


## sw/BUILD/O2-full-system-test-latest/log
Detected critical problem in logfile reco_ASYNC.log
reco_ASYNC.log:[67706:ctp-entropy-decoder]: [08:15:55][ERROR] Exception while running: InputRecord::get: no input with binding ctpconfig found. Available inputs: ctf_CTP, ctfdict_CTP, trigoffset. Rethrowing.
reco_ASYNC.log-[67706:ctp-entropy-decoder]: [08:15:55][FATAL] Unhandled o2::framework::runtime_error reached the top of main of o2-ctf-reader-workflow, device shutting down. Reason: InputRecord::get: no input with binding ctpconfig found. Available inputs: ctf_CTP, ctfdict_CTP, trigoffset
[67706:ctp-entropy-decoder]: [08:15:55][ERROR] Exception while running: InputRecord::get: no input with binding ctpconfig found. Available inputs: ctf_CTP, ctfdict_CTP, trigoffset. Rethrowing.
[67706:ctp-entropy-decoder]: [08:15:55][FATAL] Unhandled o2::framework::runtime_error reached the top of main of o2-ctf-reader-workflow, device shutting down. Reason: InputRecord::get: no input with binding ctpconfig found. Available inputs: ctf_CTP, ctfdict_CTP, trigoffset
[ERROR] Workflow crashed - PID 67706 (ctp-entropy-decoder) did not exit correctly however it's not clear why. Exit code forced to 128.


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/8d2c25513627b95731c00de6f69ff2fc8b4d21bd/slc8_x86-64/o2checkcode/1.0-local590/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13682/0de363f5596be5488cdad3cd3fa58172dd51c23a/build_O2_fullCI/pretty.html).

lietava committed:
fix: getting ctpconfig

lietava committed:
clang

lietava committed:
fix: removing std:;cout

lietava committed:
clang

lietava commented:
HI @shahor02 , I think this went through ?
Or we wait for finishing jira discussion ?

lietava committed:
removing macro's modification to have cleaner PR

lietava committed:
dec: two latency vars, rew-decoder accessing ccdb

lietava committed:
clang

lietava commented:
Hi @shahor02 , I did changes.
I modified also original name in TriggerOffset which is probably not ok ?
Then I am not sure how TriggerOffset reads database - I do not do it explicily ?

lietava commented:
@shahor02 : ok I see your answer on jira - going to change

shahor02 commented:
What you can do is: leave and use these changes but leave also the old L0_L1 data member.

shahor02 commented:
To load it from the CCDB you should do the same as in the entropy decoders, i.e. add to your raw decoder 
https://github.com/AliceO2Group/AliceO2/blob/ce63ff8c1c90453fa45b8b97eb03c02075c70469/Detectors/CTP/workflow/src/EntropyDecoderSpec.cxx#L90
and read the input with mTrigOffsBinding="trigoffset" as in https://github.com/AliceO2Group/AliceO2/blob/ce63ff8c1c90453fa45b8b97eb03c02075c70469/Detectors/Base/src/CTFCoderBase.cxx#L54.
That's all.

lietava committed:
TriggerParams old variable same name

lietava committed:
clang

lietava committed:
dev: reading of CCDB offset paraams in rawdatadecoder

lietava committed:
clang

lietava commented:
@shahor02 : can you, please, if I understood  you correctly ?

shahor02 commented on Detectors/CTP/reconstruction/src/RawDataDecoder.cxx:
```diff
@@ -89,7 +89,7 @@ int RawDataDecoder::addCTPDigit(uint32_t linkCRU, uint32_t orbit, gbtword80_t& d
   } else if (linkCRU == o2::ctp::GBTLinkIDClassRec) {
     int32_t BCShiftCorrection = -o2::ctp::TriggerOffsetsParam::Instance().customOffset[o2::detectors::DetID::CTP];
-    int32_t offset = BCShiftCorrection + o2::ctp::TriggerOffsetsParam::Instance().LM_L0 + o2::ctp::TriggerOffsetsParam::Instance().L0_L1 - 1;
+    int32_t offset = BCShiftCorrection + o2::ctp::TriggerOffsetsParam::Instance().LM_L0 + o2::ctp::TriggerOffsetsParam::Instance().L0_L1_classes - 1;
```
You have already `L0_L1_classes` as 280, do you need to subtract 1?

lietava replied:
It was like that before: there was '-1' with 280 always for classes. So I keep it backward compatible.

shahor02 replied:
@lietava  what about this?

lietava replied:
sorry, what do you mean - I commented above:
It was like that before: there was '-1' with 280 always for classes. So I keep it backward compatible.
And as L0_L1_classes is not in old database onjects - it will be 280 in that case as before (default in old TriggerOffsetParams). For new CCDB we should put also 280. Or we actually do not need new entry as with the current
L0_L1 classes is default of class and L0_L1 inputs is 281.
Or let me know if something not correct.

lietava replied:
HI @shahor02 ,
I did validation 559672 and 559243.
Validation means that o2-ctp-reco-workflow  --no-lumi --ctpinputs-decoding
gives that same digits as CTF decoder
and both are consistent wirh filling scheme.

shahor02 replied:
@lietava yes, you have ignored the comment above, could you please check?

shahor02 commented on Detectors/CTP/workflow/src/RawDecoderSpec.cxx:
```diff
@@ -135,6 +137,13 @@ void RawDecoderSpec::run(framework::ProcessingContext& ctx)
   if (fatal_flag) {
     ret = mDecoder.decodeRawFatal(inputs, filter);
   } else {
+    if (mDecodeinputs) {
+      const auto ctpcfg = inputs.get<o2::ctp::CTPConfiguration*>("ctpconfig");
+      if (ctpcfg != nullptr) {
+        mDecoder.setCTPConfig(*ctpcfg);
+      }
+    }
```
This should be done only once in the beginning of the run processing, right? Why don't you do thin in after https://github.com/AliceO2Group/AliceO2/pull/13682/files#diff-730222aede43b6dd828e45b24b01ba97b3d88fc8c073a79837364b2fb9408b65R237 ?

lietava replied:
done, also for EntropyDecoderSpec

lietava committed:
dev: config done only if run changed

lietava commented:
HI @shahor02 , I did modification for updating ctpconfig only if run changes in both ctp and ctf decoders.

alibuild commented:
Error while checking build/O2/fullCI for f8c8b975bfeb475b549b5a9f6a9b155556b53f81 at 2024-11-27 23:07:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13682/f8c8b975bfeb475b549b5a9f6a9b155556b53f81/build_O2_fullCI/pretty.html).

lietava commented:
HI @shahor02 , please, let me know if I missed something or you have more comments.

shahor02 commented:
Would not it be now 279? `L0_L1_classes` is 280.

lietava commented:
ON te current version, i.e. like in branch dev:
in the TriggerOssetsParams there is **280**:
https://github.com/AliceO2Group/AliceO2/blob/74c640eedba0f45814d421920b1aee2f9663c4ea/DataFormats/Detectors/CTP/include/DataFormatsCTP/TriggerOffsetsParam.h#L27
This was used for classes in ctprawdecoder:
https://github.com/AliceO2Group/AliceO2/blob/74c640eedba0f45814d421920b1aee2f9663c4ea/Detectors/CTP/reconstruction/src/RawDataDecoder.cxx#L92
as there was **no CCDB reading**
In CTF decoder classes are never done. There in:
https://github.com/AliceO2Group/AliceO2/blob/74c640eedba0f45814d421920b1aee2f9663c4ea/Detectors/CTP/reconstruction/include/CTPReconstruction/CTFCoder.h#L219
**281** was used in:
https://github.com/AliceO2Group/AliceO2/blob/74c640eedba0f45814d421920b1aee2f9663c4ea/Detectors/CTP/reconstruction/src/RawDataDecoder.cxx#L532
**because CCDB was read**

lietava commented:
I believe my new code is doing exactly the same and it also does correct ctpinputs decodong when done
in ctprawdecoder:
workflow --shm-segment-size 12000000000  --no-lumi --ctpinputs-decoding -b
which was not case before.
